js/graphql.md

```diff
@@ -0,0 +1,1374 @@
+connection heavy.
+
+TODO: Finish docs
+
+
+### @versioned
+
+The `@versioned` directive adds object versioning and conflict resolution to a type.
+
+#### Definition
+
+```
+directive @versioned(versionField: String = "version", versionInput: String = "expectedVersion") on OBJECT
+```
+
+#### Usage
+
+Annotate a `@model` type with the `@versioned` directive to add object versioning and conflict detection to a type.
+
+```
+type Post @model @versioned {
+  id: ID!
+  title: String!
+  version: Int!   # <- If not provided, it is added for you.
+}
+```
+
+**Creating a Post automatically sets the version to 1**
+
+```
+mutation Create {
+  createPost(input:{
+    title:"Conflict detection in the cloud!"
+  }) {
+    id
+    title
+    version  # will be 1
+  }
+}
+```
+
+**Updating a Post requires passing the "expectedVersion" which is the object's last saved version**
+
+> Note: When updating an object, the version number will automatically increment.
+
+```
+mutation Update($postId: ID!) {
+  updatePost(
+    input:{
+      id: $postId,
+      title: "Conflict detection in the cloud is great!",
+      expectedVersion: 1
+    }
+  ) {
+    id
+    title
+    version # will be 2
+  }
+}
+```
+
+**Deleting a Post requires passing the "expectedVersion" which is the object's last saved version**
+
+```
+mutation Delete($postId: ID!) {
+  deletePost(
+    input: {
+      id: $postId,
+      expectedVersion: 2
+    }
+  ) {
+    id
+    title
+    version
+  }
+}
+```
+
+Update and delete operations will fail if the **expectedVersion** does not match the version
+stored in DynamoDB. You may change the default name of the version field on the type as well as the name
+of the input field via the **versionField** and **versionInput** arguments on the `@versioned` directive.
+
+#### Generates
+
+The `@versioned` directive manipulates resolver mapping templates and will store a `version` field in versioned objects.
+
+### @searchable
+
+The `@searchable` directive handles streaming the data of an `@model` object type to
+Amazon Elasticsearch Service and configures search resolvers that search that information.
+
+> Note: Support for adding the `@searchable` directive does not yet provide automatic indexing for any existing data to Elasticsearch. View the feature request [here](https://github.com/aws-amplify/amplify-cli/issues/98).
+
+#### Definition
+
+```
+# Streams data from DynamoDB to Elasticsearch and exposes search capabilities.
+directive @searchable(queries: SearchableQueryMap) on OBJECT
+input SearchableQueryMap { search: String }
+```
+
+#### Usage
+
+Store posts in DynamoDB and automatically stream them to ElasticSearch
+via lambda and connect a searchQueryField resolver.
+
+```
+type Post @model @searchable {
+  id: ID!
+  title: String!
+  createdAt: String!
+  updatedAt: String!
+  upvotes: Int
+}
+```
+
+You may then create objects in DynamoDB that will automatically streamed to lambda
```

Comment:
"that will **be** automatically streamed"

Reply:
+1